server/pulp/plugins/rsync/configuration.py

```diff
@@ -0,0 +1,151 @@
+    """
+    def __call__(self, value, config):
+        if value:
+            return (True, None)
+        else:
+            return (False, self._err(value))
+
+    def _err(self, value):
+        return "attribute cannot be empty"
+
+
+class TypeValidation(object):
+    """
+    Validates that the value is one of allowed types
+    """
+    def __init__(self, allowed_types):
+        self.allowed_types = allowed_types
+
+    def __call__(self, value, config):
+        if [True for _type in self.allowed_types if issubclass(type(value), _type)]:
+            return (True, None)
+        else:
+            return (False, self._err(value))
+
+    def _err(self, value):
+        str_types = ", ".join(self.allowed_types)
+        return "type of atribute(%s) is not in allowed types: %s" % (type(value), str_types)
+
+
+class RequireOptionalIf(object):
+    """
+    Validates that if a particular config is present, other configs that are needed for it are also
+    present.
+    """
+    def __init__(self, required_for_attr, condition):
+        self.required_for_attr = required_for_attr
+        self.condition = condition
+
+    def __call__(self, value, config):
+        subconfig = config
+        path = []
+        if not self.condition(value):
+            return (True, None)
+
+        if isinstance(self.required_for_attr, list):
+            fifo = [(x, subconfig, path) for x in self.required_for_attr]
+        elif isinstance(self.required_for_attr, dict):
+            fifo = [(val, subconfig.get(key), [key])
+                    for key, val in self.required_for_attr.iteritems()]
+        while fifo:
+            (required_for_attr, subconfig, path) = fifo.pop(0)
+            if isinstance(required_for_attr, list):
+                for x in required_for_attr:
+                    fifo.insert(0, (x, subconfig, path))
+            elif isinstance(required_for_attr, dict):
+                for key, val in required_for_attr.iteritems():
+                    fifo.insert(0, (val, subconfig.get(key)), path + [key])
+            elif isinstance(required_for_attr, basestring):
+                if required_for_attr not in subconfig:
+                    return (False, self._err(path + [required_for_attr]))
+        return (True, None)
+
+    def _err(self, path):
+        return "%s attribute is required" % ("::".join(path))
+
+
+REMOTE_MANDATORY_KEYS = {
+    "auth_type": [OneOfValidation(["publickey", "password"]),
+                  RequireOptionalIf({"remote": ["ssh_password", "ssh_login"]},
+                                    lambda x: x == "password"),
+                  RequireOptionalIf({"remote": ["key_path", "ssh_login"]},
+                                    lambda x: x == "publickey")],
+    "host": [TypeValidation([basestring]), NonEmptyValidation()],
+    "root": [TypeValidation([basestring]), NonEmptyValidation()]
+}
+
+REMOTE_OPTIONAL_KEYS = {
+    "relative_dir": [TypeValidation([basestring]), NonEmptyValidation()],
```

Comment:
Should this be ``remote_rel_prefix`` to match what the code appears to be using?

Whatever it's called, given a comment below, it would be valuable to validate that it does not begin with a forward slash.